avi2swf/avi2swf.1
.TH avi2swf "1" "February 2006" "avi2swf" "swftools"
.SH NAME
avi2swf - Converts an avi video file to SWF.

.SH Synopsis
.B avi2swf file.avi [-o output.swf]

.SH DESCRIPTION
Converts an AVI video file to SWF.
.PP
From Version 6 on, SWF supports h.263 video rendering. This tool generates,
among some other formats (see below), SWF movies which contain such h.263 video 
from AVI files.

.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR 
    Print help and exit
.TP
\fB\-o\fR, \fB\-\-output\fR filename
    Specify output filename
.TP
\fB\-A\fR, \fB\-\-adjust\fR seconds
    Audio adjust: Shift sound -seconds to the future or +seconds into the past.
.TP
\fB\-n\fR, \fB\-\-num\fR frames
    Number of frames to encode
.TP
\fB\-m\fR, \fB\-\-mp3-bitrate\fR \fIkbps\fR
    Set the mp3 bitrate to encode audio with
.TP
\fB\-r\fR, \fB\-\-mp3-samplerate\fR \fIhz\fR
    Set the mp3 samplerate to encode audio with (default: 11025)
.TP
\fB\-s\fR, \fB\-\-scale\fR \fIval\fR
    Scale down to factor \fIval\fR. (in %, e.g. 100 = original size)
.TP
\fB\-S\fR, \fB\-\-skipframes\fR \fInum\fR
    Skip \fInum\fR frames before starting the conversion.
.TP
\fB\-p\fR, \fB\-\-flip\fR 
    Turn movie upside down
.TP
\fB\-q\fR, \fB\-\-quality\fR \fIval\fR
    Set the quality to \fIval\fR. (0-100, 0=worst, 100=best, default:80)
.TP
\fB\-k\fR, \fB\-\-keyframe\fR 
    Set the number of intermediate frames between keyframes.
.TP
\fB\-x\fR, \fB\-\-extragood\fR 
    Enable some \fIvery\fR expensive compression strategies. You may
    want to let this run overnight.
.TP
\fB\-T\fR, \fB\-\-flashversion\fR \fIn\fR
    Set output flash version to \fIn\fR. Notice: H.263 compression will only be
    used for n >= 6.
.TP
\fB\-V\fR, \fB\-\-version\fR 
    Print program version and exit


ChangeLog
0.0.1: initial release
0.0.2: 
	* pdf2swf: Added support for pdf bitmaps, clipping, state saving/restoring, colors,
	           solid fill and line width. The output SWF files are now much smaller.
	* swfcombine: Added stacking option (-t)
	* swfcombine: Small bugfixes with the command line options
	* rfxswf-lib: some bugfixes concerning sprites, cgi, shapes, jpegs.
	* swfstrings: New: Allows extraction of text from swf files
	* swfdump: New: Dumps swf file information. (And does some error checking on the SWF)
	* jpeg2swf: New: Converts one or more jpegs to a swf movie.

0.1.0: 
	* configure: Now works without automake, automatically disables 
			jpeg2swf and pdf2swf if jpeglib is not found (disable
			both) or t1lib is not found (disable pdf2swf)
	* swfcombine: bugfix: DEFINETEXT tags didn't get relocated right
	* swfcombine: bugfix: no longer builds "unstreamable" SWFs
	* rfxswf-lib: actionscript
	* rfxswf-lib: lossless image formats
	* pdf2swf: conversion of lossless image formats
	* pdf2swf: picture cache to prevent images getting stored more than once
	* pdf2swf: conversion of (hyper)links
	* swfdump: -a, -e and -XY options
	* lib/example/zlibtest: new file, demonstrates using lossless images with rfxswf-lib
	* lib/example/dumpfont: utility for dumping fonts out of swf (and generate c sourcecode for creating them)
	* lib/example/demofont: example font, generated with lib/example/dumpfont.

0.1.1:  
	* swfcombine: Wasn't able to handle converted pdfs with links. Fixed.
	* swfcombine: Introduce --dummy, -X,-Y,-r options
	* swfdump: new: -r option
	* pdf2swf: small bugfixes
	* HOWTO_pdf2swf: added section about correcting size and framerate

0.1.2:
	* swfdump: can dump all actionscript flavours (button, button2, standalone) now
	* rfxswf-lib: improved font handling
	* lib/example/makefonts.cc: new file, converts the standard fonts to swf
	* lib/example/glyphshape.c: new file, demonstrates using font characters as shapes
	* pdfswf: many bugfixes. The "pdfscreen" manual now get's converted correctly
	* rfxswf-lib: bugfix: Couldn't handle jpeg grayscale images

0.2.0:	
	* pdf2swf: TrueType support added. (using ttf2pt1)
	* rfxswf-lib: performs conversion of jpeg colorspaces now. (Makes pdf2swf cope
	               with more types of jpegs)
	* swfextract: new tool, allows to extract single objects, frames etc. from movieclips
	* swfcombine: allows ids instead of names, now, too.
	* swfdump: added text displaying functionality. (like swfstrings)
	* swfdump: now dumps also the undocumented placeobject2 actioncode.
	* lib/example/avi2swf.cc: For now, converts small (!) avi movie files to swf.  (using 
			          libavifile)

0.2.1:	
	* configure: Does checking for byte order and command substitution capability now.
	* rfxswf: Applied Anjo Krank's patch for MacOS X.
	* pdf2swf: Fixed font transformation matrix
	* swfcombine: new options: --cat, --stack1, --merge, --frame
	* swfs/: new directory, holds some standard SWFs primarily used for pdf2swf
	         installs into ${DATADIR}/swftools/swfs
	* pdf2swf: New options --viewer, --preloader, -b, -l
	* pdf2swf: Fonts are now in ${DATADIR}/swftools/fonts

0.2.2:	
	* swfs/: Added/Wrote sources for some of the swf files 
	* swfdump: Major bugfix in placeobject2, hex output function
	* swfextract: Can now extract png, jpeg and mp3.
	* swfcombine: big endian fixes
	* swfextract: big endian fixes
	* pdf2swf: Changed from 100dpi to 72dpi conversion
	* pdf2swf: Some optimizations, fix the zero string problem.
	* lib/: Added Bladeenc 0.94.1 for mp3 compression
	* lib/example/avi2swf.cc: Can now handle big files. Slightly better compression ratio. Sound.
	* lib/example/sound.c: New file, demonstrates use of the sound module in librfxswf.

0.2.3: 
	* pdf2swf: New option --fontdir, can now handle external fonts.
	* lib/modules/swftext.c: Several font-extraction bugfixes
	* src/swfdump: Bugfixes, ids and depths are now in decimal
	* src/swfcombine: Bugfixes.
	* configure.in: Cleaned up.

0.3.0:
	* pdf2swf: The --ignore option now actually makes the pdf files smaller.
	* rfxswf: Can now write DefineFont2 tags
	* swfextract/swfcombine: Bugfixes (DefineEditText)
	* swfdump: New option --used, prints out referred IDs of each tag.
	* rfxswf: Can now read and write Flash 6 (MX) compressed SWFs. (SWCs)
	* pdf2swf: Added compression flag (-z)
	* swfcombine: Added compression flag (-z)
	* configure.in: zlib is now required.
	* src/png2swf: New tool.
	* swfdump: Prints some image infos now.
	* pdf2swf: Fixed Image handling
	* avi2swf/: Is part of the main distribution now
	* src/wav2swf: New tool.

0.3.1:
	* configure: Didn't create all Makefiles when t1lib or jpeglib were absent. Fixed.
	* swfs/Makefile: FreeBSD fixes.

0.4.0:
	* pdf2swf/xpdf: Upgraded PDF Parser to Version 1.01.
	* pdf2swf: Can now handle Type 3 Fonts. (As Bitmaps)
	* src/swfcombine: Completely rewritten, using the rfxswf routines
	* configure: Check for libtool
	* lib/rfxswf: Fixed some parsing bugs (swffont, swfdump)
	* pdf2swf: Added antialiasing capabilities for Type 3 Fonts
	* lib/examples/makefonts: Fonts: Bounding boxes are now handled correctly
	* png2swf: Support for more image types
	* pdf2swf: New option -t (--stop)

0.4.1:
	* src/swfcombine: Bugfix with -x option
	* lib/Makefile: MacOSX fixes
	* pdf2swf/SWFOutput.cc: fixed undefined sqrt()
	* src/png2swf: Support for transparency
	* pdf2swf: Support for CID Fonts
	* src/swfcombine: Bugfixes
	* lib/rfxswf: Added Sprite Unfolding routines

0.4.2:
	* src/Makefile: OpenBSD fixes
	* lib/lame: Removed *.o files from archive
	* avi2swf: Implemented --flip option
	* pdf2swf: Implemented --flashversion option
	* pdf2swf: Implemented Viewer gadgets- it's now possible to call Flash functions via hyperlink

0.4.3:
	* src/jpeg2swf: Grayscale fix
	* lib/rfxswf: Improved MX support
	* pdf2swf: Command line fixes
	* swfdump: Added -f (--frames) option
	* png2swf: Added support for mode 6 images
	* swfs: Support for print protection in simple_viewer

avi2swf/avi2swf.cc
/* avi2swf.cc
   Convert avi movie files into swf.

   Part of the swftools package.
   
   Copyright (c) 2001,2002,2003 Matthias Kramm <kramm@quiss.org>
 
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>

#include "../config.h"

#include "../lib/args.h"
#include "v2swf.h"
#ifdef WIN32
#include "videoreader_vfw.hh"
#else
#include "videoreader_avifile.hh"
#endif

static char * filename = 0;
static char * outputfilename = "output.swf";
int verbose = 0;

static int quality = 80;
static double scale = 1.0;
static int flip = 0;
static int expensive = 0;
static int flashversion = 6;
static int keyframe_interval = -1;
static int skip = 0;
static float audio_adjust = 0;
static int mp3_bitrate = 32;
static int samplerate = 11025;
static int numframes = 0;

static struct options_t options[] = {
{"h", "help"},
{"o", "output"},
{"A", "adjust"},
{"n", "num"},
{"m", "mp3-bitrate"},
{"r", "mp3-samplerate"},
{"s", "scale"},
{"S", "skipframes"},
{"p", "flip"},
{"q", "quality"},
{"k", "keyframe"},
{"x", "extragood"},
{"T", "flashversion"},
{"V", "version"},
{0,0}
};

int args_callback_option(char*name,char*val)
{
    if(!strcmp(name, "V")) {
        printf("avi2swf-ng - part of %s %s\n", PACKAGE, VERSION);
        exit(0);
    } 
    else if(!strcmp(name, "o")) {
	outputfilename = val;
	return 1;
    }
    else if(!strcmp(name, "n")) {
	numframes = atoi(val);
	return 1;
    }
    else if(!strcmp(name, "d")) {
	scale = atof(val);
	return 1;
    }
    else if(!strcmp(name, "q")) {
	quality = atoi(val);
	if(quality<0)
	    quality = 0;
	if(quality>100)
	    quality = 100;
	return 1;
    }
    else if(!strcmp(name, "p")) {
	flip = 1;
	return 0;
    }
    else if(!strcmp(name, "k")) {
	keyframe_interval = atoi(val);
	return 1;
    }
    else if(!strcmp(name, "A")) {
	audio_adjust = atof(val);
	return 1;
    }
    else if(!strcmp(name, "v")) {
	verbose = 1;
	return 0;
    }
    else if(!strcmp(name, "T")) {
	flashversion = atoi(val);
	return 1;
    }
    else if(!strcmp(name, "x")) {
	expensive = 1;
	return 0;
    }
    else if(!strcmp(name, "m")) {
	mp3_bitrate = atoi(val);
	return 1;
    }
    else if(!strcmp(name, "r")) {
        samplerate = atoi(val);
        if(samplerate >= 11000 && samplerate <= 12000)
            samplerate = 11025;
        else if(samplerate >= 22000 && samplerate <= 23000)
            samplerate = 22050;
        else if(samplerate >= 44000 && samplerate <= 45000)
            samplerate = 44100;
        else {
            fprintf(stderr, "Invalid samplerate: %d\n", samplerate);
            fprintf(stderr, "Allowed values: 11025, 22050, 44100\n", samplerate);
            exit(1);
        }
        return 1;
    }
    else if(!strcmp(name, "S")) {
	skip = atoi(val);
	return 1;
    }
    else if(!strcmp(name, "s")) {
	scale = atoi(val)/100.0;
	if(scale>1.0 || scale<=0) {
	    fprintf(stderr, "Scale must be in the range 1-100!\n");
	    exit(1);
	}
	return 1;
    }
    fprintf(stderr, "Unknown option: -%s\n", name);
    exit(1);
}
int args_callback_longoption(char*name,char*val)
{
    return args_long2shortoption(options, name, val);
}
void args_callback_usage(char *name)
{
    printf("\n");
    printf("Usage: %s file.avi [-o output.swf]\n", name);
    printf("\n");
    printf("-h , --help                    Print help and exit\n");
    printf("-o , --output filename         Specify output filename\n");
    printf("-A , --adjust seconds          Audio adjust: Shift sound -seconds to the future or +seconds into the past.\n");
    printf("-n , --num frames              Number of frames to encode\n");
    printf("-m , --mp3-bitrate <kbps>      Set the mp3 bitrate to encode audio with\n");
    printf("-r , --mp3-samplerate <hz>     Set the mp3 samplerate to encode audio with (default: 11025)\n");
    printf("-s , --scale <val>             Scale down to factor <val>. (in %, e.g. 100 = original size)\n");
    printf("-S , --skipframes <num>        Skip <num> frames before starting the conversion.\n");
    printf("-p , --flip                    Turn movie upside down\n");
    printf("-q , --quality <val>           Set the quality to <val>. (0-100, 0=worst, 100=best, default:80)\n");
    printf("-k , --keyframe                Set the number of intermediate frames between keyframes.\n");
    printf("-x , --extragood               Enable some *very* expensive compression strategies.\n");
    printf("-T , --flashversion <n>        Set output flash version to <n>.\n");
    printf("-V , --version                 Print program version and exit\n");
    printf("\n");
}
int args_callback_command(char*name,char*val)
{
    if(filename) {
        fprintf(stderr, "Only one file allowed. You supplied at least two. (%s and %s)\n",
                 filename, name);
    }
    filename = name;
    return 0;
}

static char toabuf[128];
static char*ftoa(double a)
{
    sprintf(toabuf, "%f", a);
    return toabuf;
}
static char*itoa(int a)
{
    sprintf(toabuf, "%d", a);
    return toabuf;
}

#ifdef DO_SIGNALS
pthread_t main_thread;
static void sigterm(int sig)
{
    if(pthread_equal (pthread_self(), main_thread))
    {
	if(frameno>0 && !shutdown_avi2swf) {
	    if(verbose)
		printf("Thread [%08x] got sigterm %d\n", pthread_self(), sig);
	    shutdown_avi2swf++;
	} else {
	    exit(1);
	}
    }
}
#endif

int main (int argc,char ** argv)
{ 
    videoreader_t video;
    v2swf_t v2swf;
    int ret;
    FILE*fi;

#ifdef DO_SIGNALS
    signal(SIGTERM, sigterm);
    signal(SIGINT , sigterm);
    signal(SIGQUIT, sigterm);
    main_thread = pthread_self();
#endif

    processargs(argc, argv);
    if(!filename) {
	fprintf(stderr, "You must supply a filename");
	exit(0);
    }
    if(keyframe_interval<0) {
	if(flashversion>=6)
	    keyframe_interval=20;
	else
	    keyframe_interval=5;
    }
    
    fi = fopen(outputfilename, "wb");
    if(!fi) {
	fflush(stdout); fflush(stderr);
	fprintf(stderr, "Couldn't open %s\n", outputfilename);
	exit(1);
    }
    
#ifdef WIN32
    ret = videoreader_vfw_open(&video, filename);
#else
    ret = videoreader_avifile_open(&video, filename);
#endif

    if(ret<0) {
	fprintf(stderr, "Error opening %s\n", filename);
	exit(1);
    }

    if(verbose) {
	printf("| video framerate: %f\n", video.fps);
	printf("| video size: %dx%d\n", video.width, video.height);
	printf("| audio rate: %d\n", video.samplerate);
	printf("| audio channels: %d\n", video.channels);
    }

    ret = v2swf_init(&v2swf, &video);
    if(verbose)
	v2swf_setparameter(&v2swf, "verbose", "1");
    if(numframes)
	v2swf_setparameter(&v2swf, "numframes", itoa(numframes));
    v2swf_setparameter(&v2swf, "quality", itoa(quality));
    v2swf_setparameter(&v2swf, "blockdiff", "0");
    v2swf_setparameter(&v2swf, "blockdiff_mode", "exact");
    v2swf_setparameter(&v2swf, "mp3_bitrate", itoa(mp3_bitrate));
    v2swf_setparameter(&v2swf, "samplerate", itoa(samplerate));
    //v2swf_setparameter(&v2swf, "fixheader", "1");
    //v2swf_setparameter(&v2swf, "framerate", "15");
    v2swf_setparameter(&v2swf, "scale", ftoa(scale));
    v2swf_setparameter(&v2swf, "prescale", "1");
    v2swf_setparameter(&v2swf, "flash_version", itoa(flashversion));
    v2swf_setparameter(&v2swf, "keyframe_interval", itoa(keyframe_interval));
    if(expensive)
	v2swf_setparameter(&v2swf, "motioncompensation", "1");
    if(flip)
	video.setparameter(&video, "flip", "1");
    if(verbose)
	video.setparameter(&video, "verbose", "1");

    if(!verbose)
	printf("\n");

    if(audio_adjust>0) {
	int num = ((int)(audio_adjust*video.samplerate))*video.channels*2;
	void*buf = malloc(num);
	video.getsamples(&video, buf, num);
	free(buf);
    } else if(audio_adjust<0) {
	int num = (int)(-audio_adjust*video.fps);
	void*buf = malloc(video.width*video.height*4);
	int t;
	for(t=0;t<num;t++) {
	    video.getimage(&video, buf);
	}
	free(buf);
    }

    if(skip) {
	int t;
	void*buf = malloc(video.width*video.height*4);
	for(t=0;t<skip;t++) {
	    video.getimage(&video, buf);
	    video.getsamples(&video, buf, (int)((video.samplerate/video.fps)*video.channels*2));
	    if(!verbose) {
		printf("\rSkipping frame %d", video.frame);fflush(stdout);
	    }
	}
	free(buf);
    }

    char buffer[4096];
    while(1) {
	int l=v2swf_read(&v2swf, buffer, 4096);
	fwrite(buffer, l, 1, fi);
	if(!l)
	    break;
	if(!verbose) {
	    printf("\rConverting frame %d", video.frame);fflush(stdout);
	}
    }
    if(!verbose)
	printf("\n");
    fclose(fi);
    v2swf_backpatch(&v2swf, outputfilename);
    v2swf_close(&v2swf);
}

